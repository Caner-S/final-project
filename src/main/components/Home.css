.businessCaseWarning {
    color: red;
}

.businessCaseBorder {
    text-align: center;
}