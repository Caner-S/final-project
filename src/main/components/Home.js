import React from 'react';
import Grid from "@material-ui/core/Grid";
import floorPlan from "../../6th Floor Holborn.svg";
import Spaces from "./Spaces";
import './Home.css';
import ToggleButton from '@material-ui/lab/ToggleButton';
import ToggleButtonGroup from '@material-ui/lab/ToggleButtonGroup';
import {getAllSpaces} from "../dao/SpaceDao";
import {getUnavailableBookings} from "../dao/BookingDao";
import TextField from "@material-ui/core/TextField";
import SingleBooking from "./SingleBooking";

class Home extends React.Component {
    constructor(props) {
        super(props);
        let today = new Date().toISOString().slice(0, 10);
        this.state = {
            spaces: [],
            todayISO: today+"T23:59",
            arrivalDate: today+"T09:00",
            departureDate: today+"T17:00",
            loadSpaces: false,
            view: 'table',
            currentSpace: undefined,
        };
    }

    async componentDidMount() {
        const allSpaces = await getAllSpaces();
        this.setState({ spaces: allSpaces });
        this.getAllAvailableSpace();
    }

    componentDidUpdate(prevProps, prevState, snapshot) {
        this.fillAvailableSpaces();

    }

    getAllAvailableSpace = async () => this.processSpaces(getUnavailableBookings(this.state.arrivalDate, this.state.departureDate, this.state.todayISO));

    processSpaces = async (newSpaces) => {
        let currentComponent = this;
        const endTimeDate = new Date(this.state.departureDate).valueOf();
        try {
            const snapshot = await newSpaces;
            snapshot.onSnapshot((snapshot) => {
                snapshot.forEach(function(doc) {
                    if(endTimeDate > doc.data().arrivalDate){
                        currentComponent.setState({spaces: currentComponent.state.spaces.filter(item => item.id !== doc.data().spaceId)});
                        currentComponent.svgColourer(doc.data().seatNumber, "none");
                    }
                });
            })
        } catch (e) {
            console.log(e)
        }
    }


    handleAlignment = (e, newAlignment) => {
        e.preventDefault();
        this.setState({view: newAlignment});
    };

    businessCaseWarning(){
        if(this.props.businessCase){
            return <div className="businessCaseBorder">
                <h3 className="businessCaseWarning">Due to the current pandemic, all requests must be approved with a suitable business case. </h3>
                <h3 className="businessCaseWarning">  The status of these requests can be found on the 'My Booking' page.</h3>
            </div>
                }
    }

    bookingType(){
        if(this.state.view === 'table'){
            return  <Grid container>

                <Grid item xs={6} s={6} zeroMinWidth>
                    <object data={floorPlan} id="svg" aria-label="floorPlan" onLoad={() => this.setState({loadSpaces: true})}/>
                </Grid>
                <Grid item xs={6} s={6} zeroMinWidth>

                    <Spaces businessCase={this.props.businessCase} spaces={this.state.spaces} arrivalDate={this.state.arrivalDate} departureDate={this.state.departureDate}/>
                </Grid>
            </Grid>

        }else if (this.state.view === 'floorplan'){
            return  <Grid container >

                <Grid item xs={8} s={6} zeroMinWidth>
                    <object data={floorPlan} id="svg" aria-label="floorPlan" onLoad={() => this.setState({loadSpaces: true})}/>
                </Grid>

                <Grid item xs={4} s={6} zeroMinWidth>
                    <SingleBooking businessCase={this.props.businessCase} space={this.state.currentSpace} arrivalDate={this.state.arrivalDate} departureDate={this.state.departureDate}/>
                </Grid>


            </Grid>

        }
    }

    updateDeparture = e => {
        const arrivalDay = this.state.arrivalDate.substring(8, 10);
        const departureDate = e.target.value.substring(8, 10);
        if(arrivalDay === departureDate) {
            this.setState({
                [e.target.name]: e.target.value
            });
        }else {
            alert("Arrival and departure must be on the same day.");

        }
    }

    updateArrival = e => {
        this.setState({
            [e.target.name]: e.target.value,
            departureDate: e.target.value,
        });

    }



    svgColourer (space, colour)  {
        let currentComponent = this;
        const a = document.getElementById("svg");
        const svgDoc = a.contentDocument;
        const elements = svgDoc.getElementsByClassName(space.toString());
        if(typeof elements[0] !== "undefined"){
            elements[0].style.fill = colour;
            elements[0].onclick = function() { currentComponent.getCurrentSelection(space);};
        }
    }

    getCurrentSelection(space){
        let obj = this.state.spaces.find(o => o.data().number === space);
        this.setState({currentSpace: obj })
    }

    fillAvailableSpaces() {
        if (this.state.loadSpaces) {
            this.state.spaces.map(space => (
                this.svgColourer(space.data().number, "green")
            ));
        }
    }


    render() {

        return (
            <div>
                <Grid container>

                    <Grid item xs={6} s={6} zeroMinWidth>
                <ToggleButtonGroup
                    value={this.state.view}
                    exclusive
                    onChange={this.handleAlignment}
                    aria-label="booking view"
                >
                    <ToggleButton value="table" aria-label="table view">
                        Table View
                    </ToggleButton>
                    <ToggleButton value="floorplan" aria-label="floorplan view">
                        Floorplan View
                    </ToggleButton>
                </ToggleButtonGroup>
                    </Grid>
                    <Grid item xs={6} s={6} zeroMinWidth>
                <form >
                    <TextField
                        id="datetime-local-arrive"
                        name="arrivalDate"
                        label="Arrival Date"
                        type="datetime-local"
                        onChange={this.updateArrival}
                        value={this.state.arrivalDate}
                        InputLabelProps={{
                            shrink: true,
                        }}
                    />
                    <TextField
                        id="datetime-local-depart"
                        name="departureDate"
                        label="Departure Date"
                        type="datetime-local"
                        onChange={this.updateDeparture}
                        value={this.state.departureDate}
                        InputLabelProps={{
                            shrink: true,
                        }}
                    />
                </form>
                    </Grid>
                </Grid>
            {this.businessCaseWarning()}
            {this.bookingType()}

            </div>
        );
    }
}
export default Home;