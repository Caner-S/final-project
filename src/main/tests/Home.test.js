import React from 'react';
import renderer from 'react-test-renderer';

import Home from '../components/Home';

it('renders correctly when there are no items', () => {
    const tree = renderer.create(<Home />).toJSON();
    expect(tree).toMatchSnapshot();
});