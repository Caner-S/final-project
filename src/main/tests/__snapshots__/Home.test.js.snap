// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly when there are no items 1`] = `
<div>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-zeroMinWidth MuiGrid-grid-xs-6"
      s={6}
    >
      <div
        aria-label="booking view"
        className="MuiToggleButtonGroup-root"
        role="group"
      >
        <button
          aria-label="table view"
          aria-pressed={true}
          className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal Mui-selected"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
          value="table"
        >
          <span
            className="MuiToggleButton-label"
          >
            Table View
          </span>
        </button>
        <button
          aria-label="floorplan view"
          aria-pressed={false}
          className="MuiButtonBase-root MuiToggleButton-root MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal"
          disabled={false}
          onBlur={[Function]}
          onChange={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
          value="floorplan"
        >
          <span
            className="MuiToggleButton-label"
          >
            Floorplan View
          </span>
        </button>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-zeroMinWidth MuiGrid-grid-xs-6"
      s={6}
    >
      <form>
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
            htmlFor="datetime-local-arrive"
            id="datetime-local-arrive-label"
          >
            Arrival Date
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="datetime-local-arrive"
              name="arrivalDate"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="datetime-local"
              value="2021-03-22T09:00"
            />
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiFormLabel-filled"
            data-shrink={true}
            htmlFor="datetime-local-depart"
            id="datetime-local-depart-label"
          >
            Departure Date
          </label>
          <div
            className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiInput-input"
              disabled={false}
              id="datetime-local-depart"
              name="departureDate"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="datetime-local"
              value="2021-03-22T17:00"
            />
          </div>
        </div>
      </form>
    </div>
  </div>
  <div
    className="MuiGrid-root MuiGrid-container"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-zeroMinWidth MuiGrid-grid-xs-6"
      s={6}
    >
      <object
        aria-label="floorPlan"
        data="6th Floor Holborn.svg"
        id="svg"
        onLoad={[Function]}
      />
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-zeroMinWidth MuiGrid-grid-xs-6"
      s={6}
    >
      <div
        className="border"
      >
        <div
          className="MuiPaper-root MuiTableContainer-root container MuiPaper-elevation1 MuiPaper-rounded"
        >
          <table
            aria-label="simple table"
            className="MuiTable-root"
            role={null}
          >
            <thead
              className="MuiTableHead-root"
              role={null}
            >
              <tr
                className="MuiTableRow-root MuiTableRow-head"
                role={null}
              >
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head"
                  scope="col"
                >
                  SeatÂ Number
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Capacity
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Type
                </th>
                <th
                  aria-sort={null}
                  className="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight"
                  scope="col"
                >
                  Floor
                </th>
              </tr>
            </thead>
            <tbody
              className="MuiTableBody-root"
              role={null}
            />
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;
